File: leetcode/1232. Check If It Is a Straight Line/main.ex
defmodule Solution do
  @spec check_straight_line(coordinates :: [[integer]]) :: boolean
  def check_straight_line(coordinates) do
    case coordinates do
      [] ->
        false

      [_] ->
        false

      [first, second | rest] ->
        first_slope = slope([first, second])

        [second | rest]
        |> Enum.chunk_every(2, 1, :discard)
        |> Enum.map(&slope(&1))
        |> Enum.all?(&is_approx(&1, first_slope, 0.0001))
    end
  end

  defp slope([[x1 | _], [x2 | _]]) when x1 == x2, do: :infinity
  defp slope([[x1, y1], [x2, y2]]), do: (y2 - y1) / (x2 - x1)

  defp is_approx(a, b, eps) when a == :infinity or b == :infinity, do: a === b
  defp is_approx(a, b, eps), do: abs(a - b) < eps
end

File: leetcode/1232. Check If It Is a Straight Line/main.rs
impl Solution {
    pub fn check_straight_line(coordinates: Vec<Vec<i32>>) -> bool {
        if coordinates.len() < 2 {
            return true;
        }
        let first_slope = Solution::slope(&coordinates[1], &coordinates[0]);
        const EPSILON: f64 = 0.01;
        for i in 2..coordinates.len() {
            let curr_slope = Solution::slope(&coordinates[i - 1], &coordinates[i]);
            if (first_slope - curr_slope).abs() > EPSILON {
                return false;
            }
        }
        true
    }

    fn slope(from: &Vec<i32>, to: &Vec<i32>) -> f64 {
        if to[0] == from[0] {
            return f64::INFINITY;
        }
        (to[1] as f64 - from[1] as f64) / (to[0] as f64 - from[0] as f64)
    }
}